### PR TITLE
Expose tfio.IOTensor class and from_audio and tfio.IOTensor.to_dataset()

### PR DESCRIPTION
**Pull request by yongtang**
This PR tries to expose a tfio.IOTensor which could be applied to and io related data which are indexable (__getitem__ and __len__)

The idea is to bind `__getitem__` and `__len__` to kernel ops in run time, so that is is not necessarily to read everything in memory.

The first file format is the WAV file. With tfio.IOTensor dtype and shape are exposed with `__getitem__` and `__len__`.

Further, a rate property has been exposed specifically for Audio/WAV file which gives sample rate.

This tfio.IOTensor only works in eager mode.

In additional this PR also converts WavDataset to use IOTensor (instead of direct C++ implementation).

This PR also carries #420.

Note as was discussed, rebatch has been dropped. Instead, a PR to core tensorflow repo will be opened.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by yongtang** [comment]
@BryanCutler I reorganized the python class as was suggested, please take a look.

**Comment by yongtang** [comment]
@BryanCutler @terrytangyuan Some note about the changes in this PR:
1) This PR adds JSONIOTensor which actually is built on Apache Arrow C++ (didn't realize Arrow already support so many formats). - We probably could take another look and see what else in Arrow could be build into tfio.
2) This PR updates KafkaDataset which keep the old API but used the C++ implementation of `KafkaIterable`. So KafkaDataset is used for iterations, and passing to tf.keras directly (if data has already been preprocessed).
3) This PR adds KafkaIOTensor and use the same C++ implementation of `KafkaIterable`. It adds a thin layer to store data in memory. This is used for indexing and slicing, and any complicated feature engineering (that could not be done with just a iterable).
4) You can convert IOTensor to dataset(), this is for people to already done heavy feature engineering such as normalize over summation, full range shuffling, etc.

One final note, is that `KafkaIterable` is about 200 lines of C++, while in comparison, the old handcrafted KafkaDataset C++ is about 450+ lines of C++. I think this is a nice code reduction.

Please take a look and see if this is OK. If it is fine, I am planning to roll out the new implementation to most of the remaining ops.

**Comment by yongtang** [comment]
@BryanCutler @terrytangyuan one final note is that, all internal implementation batches and caches a large chunk automatically so I would assume there will be a slight improvement in performance. This is especially the case for non-image files where each `element` is very small (such as 4 bytes for an integer).

File: tensorflow_io/json/kernels/json_kernels.cc [hunk 1]
```diff
@@ -199,5 +205,229 @@ REGISTER_KERNEL_BUILDER(Name("ReadJSON").Device(DEVICE_CPU),
...
+    }
+
+    dtypes_.clear();
+    shapes_.clear();
+    columns_.clear();
+    for (size_t i = 0; i < columns_index_.size(); i++) {
+      int column_index = columns_index_[i];
+      ::tensorflow::DataType dtype;
+      switch (table_->column(column_index)->type()->id()) {
+      case ::arrow::Type::BOOL:
```
Comment (BryanCutler) [inline tensorflow_io/json/kernels/json_kernels.cc]: could you use `arrow::adapters::tensorflow::GetTensorFlowType` here?

**Comment by jiachengxu** [comment]
Hi @yongtang, It is so great to have `from_json`! Here are some of my thoughts:
- As you mentioned in #420 and also in this implementation of `from_json`, arrow is a good fit to handle splittable JSON(ndjson).  I am thinking about if I should also switch to use arrow for the `list_json_columns` and `read_json` ops. Since the arrow uses rapidjson underneath, and according to some experiments https://github.com/mloskot/json_benchmark, maybe arrow cound give better performance. 
- The pure JSON is kind of special, it is not splittable and indexible, so I am thinking about that maybe if it is impossible to implement `from_json` for pure JSON because I think current `from_json` is `from_ndjson` indeed.

File: tensorflow_io/core/python/ops/kafka_dataset_ops.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      metadata = [e for e in configuration or []]
+      if servers is not None:
+        metadata.append("bootstrap.servers=%s" % servers)
+      resource, _, _ = core_ops.kafka_iterable_init(
+          subscription, metadata=metadata,
+          container=scope,
+          shared_name="%s/%s" % (subscription, uuid.uuid4().hex))
+
+      capacity = 4096
+      dataset = tf.compat.v2.data.Dataset.range(0, sys.maxsize, capacity)
```
[the remark this answers is not in the document]
Comment (yongtang) [reply tensorflow_io/core/python/ops/kafka_dataset_ops.py]: @BryanCutler Yes it could be easily adjustable, and it could even be a 1-d array (than a constant). Added an issue https://github.com/tensorflow/io/issues/445 for that. Will try to write an example once I find some time.

File: tensorflow_io/core/python/ops/kafka_io_tensor_ops.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+               internal=False):
+    with tf.name_scope("KafkaIOTensor") as scope:
+      metadata = [e for e in configuration or []]
+      if servers is not None:
+        metadata.append("bootstrap.servers=%s" % servers)
+      resource, dtypes, shapes = core_ops.kafka_indexable_init(
+          subscription, metadata=metadata,
+          container=scope,
+          shared_name="%s/%s" % (subscription, uuid.uuid4().hex))
+      print("VVV: ", dtypes, shapes)
```
[the remark this answers is not in the document]
Comment (yongtang) [reply tensorflow_io/core/python/ops/kafka_io_tensor_ops.py]: @BryanCutler Thanks. Removed.

File: tensorflow_io/json/kernels/json_kernels.cc (hunk 1 above)
Comment (yongtang) [reply tensorflow_io/json/kernels/json_kernels.cc]: @BryanCutler It is a little complicated as GetTensorFlowType is in a header file in arrow library. So directly include the header in two `.cc` files will not work. I created a wrapper instead to avoid linking issues.

File: tensorflow_io/core/python/ops/io_tensor_ops.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+
+    Args:
+
+    Returns:
+      A `tf.data.Dataset` with value obtained from this `IOTensor`.
+    """
+    return _BaseIOTensorDataset(
+        self.spec, self._resource, self._function)
+
+class _ColumnIOTensor(_BaseIOTensor):
```
[the remark this answers is not in the document]
Comment (yongtang) [reply tensorflow_io/core/python/ops/io_tensor_ops.py]: @BryanCutler `ColumnIOTensor` is essentially a `one datatype single tensor/array`. 

I could not find a better name. Maybe some suggestions?

See some of the discussions in https://github.com/tensorflow/io/issues/315#issuecomment-524568208

**Comment by yongtang** [comment]
@BryanCutler I plan to merge this PR shortly. It may not be perfect through I think we could just move forward and polish in follow up PRs (might be many). 

Created one issue #445 to track capacity and translation of batch into an array of capacities (instead of one constant).

Also added a comment in https://github.com/tensorflow/io/issues/315#issuecomment-524568208 to expand the discussion.

**Comment by yongtang** [comment]
@jiachengxu I think `list_json_columns` could be deprecated, as it meant to be a workaround to give user an easy way to check for columns. Instead it could be integrated into IOTensor and it automatically gives you columns (and more meta data).

The json and ndjson could be consolidated with one flag passed to `from_json` to control if root element is to be parse or not. (ndjson does not have root element, json does have one root element).
